transport-classes-io_uring/src/main/java/io/netty/channel/uring/AbstractIoUringServerChannel.java

```diff
@@ -120,9 +122,26 @@ protected int scheduleRead0(boolean first) {
 
             int fd = fd().intValue();
             IoUringIoRegistration registration = registration();
+
+            // Depending on if socketWasEmpty is true we will arm the poll upfront and skip the initial transfer
+            // attempt.
+            // See https://github.com/axboe/liburing/wiki/io_uring-and-networking-in-2023#socket-state
+            //
+            // Depending on if this is the first read or not we will use Native.IORING_ACCEPT_DONT_WAIT.
+            // The idea is that if the socket is blocking we can do the first read in a blocking fashion
+            // and so not need to also register POLLIN. As we can not 100 % sure if reads after the first will
+            // be possible directly we schedule these with Native.IORING_ACCEPT_DONT_WAIT. This allows us to still be
+            // able to signal the fireChannelReadComplete() in a timely manner and be consistent with other
+            // transports.
+            final short ioPrio;
+
+            if (first) {
+                ioPrio = socketWasEmpty ? Native.IORING_RECVSEND_POLL_FIRST : 0;
```

Comment:
I think it'd be easier to follow if we extracted a method for this so `socketWasEmpty` becomes a parameter, instead of a field. It's effectively a secret parameter right now, where we have a finally-clause to reset it, for instance. With the field, we have to worry about reentrancy and managing the state between calls.

Reply:
I see it'll have to be channelled through `scheduleRead` because this is `scheduleRead0`. I think it'd be worth it to pull this up since `AbstractionIoUringStreamChannel` has the same needs and suffer the same complexity.

Reply:
You mean changing the methods to take an extra param that indicates that there is no data on the socket when it is called ?

Reply:
Yeah, exactly.

Reply:
@chrisvest PTAL again